### ISSUE TITLE
Abnormal memory spikes in .NET Core 2.1 when running on containers with memory limit set

### ISSUE THREAD

**Issue by passuied** [dotnet/corefx#33346]
Hi,

We have been investigating some weird memory issues on docker containers running aspnet core 2.1.5. I have been using Sebastien Ros' [memory leak](https://github.com/sebastienros/memoryleak) project (specifically bigstring endpoint) as benchmark...
What we're finding is that in .NET core 2.1, when constraining container memory to 256MB or less and applying some load, the memory quickly jumps up until the container dies.
If only limiting to 512MB, the memory jumps all the way to 480MB, then drops to 280MB and stays stable...

Changing `<ServerGarbageCollection>`  to false solves the problem but given the fact it seems the surge only happens initially, it seems something should be fixed at .NET Core level to make sure it respects the memory limit set.

Please also note that if we are also constraining the CPU, the memory stays under control (around 130MB). Memory limit therefore seems linked to the number of CPUs available (as indicated in https://blog.markvincze.com/troubleshooting-high-memory-usage-with-asp-net-core-on-kubernetes/)

Finally, please note that when running the same service on .NET Core 2.0, the memory only jumps to 200MB (80%) ... so this behavior seems to be new in 2.1

Is this a known issue / expected behavior?

Thanks,
Patrick

**Comment by danmosemsft** [comment]
What is your configuration?

Maybe this? 
https://github.com/dotnet/coreclr/issues/19060

**Comment by passuied** [comment]
Could be. They mention a fix in 2.1 back in August though. I'm using the latest 2.1.5 image though...

**Comment by passuied** [comment]
@danmosemsft. I have just forked @sebastienros memoryleak repo and dockerized it:
https://github.com/passuied/memoryleak/tree/feature/containerize-service

Running it on a t3.2xlarge EC2 instance.

Running it in docker with 256MB memory limit:
```
$ docker run -e ASPNETCORE_URLS=http://*:8080 -e CSOD_EnvironmentName=dev-passuied -t -i --rm --memory 256m -p8080:8080 memoryleak:2.1.5 dotnet MemoryLeak.dll
```
Applying load via jmeter (5 concurrent users) on /api/bigstring endpoint
This leads to the container running out of memory after only 1 min or so...

If limiting cpu as follows:

**Comment by danmosemsft** [comment]
@janvorli does this seem familiar?

**Comment by janvorli** [comment]
@danmosemsft no, but I assume that the smaller number of available CPUs probably means lower parallelism in processing of the requests and thus possibly lower memory consumption / garbage generation. And it also means lower number of GC heaps that can also have effect on how often the GC is performed.
CC: @Maoni0

**Comment by sebastienros** [comment]
FYI this endpoint that is hit is turning a new 10K chars String instance on each call.

I will do some testing on how Docker memory settings are impacting perf. Including memory.

@passuied can try to disable server GC? It's explained on my repo. With this low amount of memory it might be more adapted. We still need to investigate.

**Comment by passuied** [comment]
@sebastienros, yes disabling server GC solves the issue.
I wasn't sure if it was a good trade off performance-wise... Seems like this spike only happens when the application starts.
If I raise the memory limit to 512MB, the memory jumps to 480MB then goes down to 200MB. Seems like the spike is only temporary...
What I would like to figure out is whether it is typically better when dealing with lower memory containers to:
- disable Server GC
- keep Server GC true but constrain CPU
- or ideally, fix what causes .NET core 2.1 to claim more memory than what's available ;)

Any pointers will be welcome

**Comment by karelz** [comment]
cc @brianrob @billwert

**Comment by sebastienros** [comment]
@passuied from my tests it's only an issue if your limit is CPU, like if you run at max CPU, but otherwise it's not a problem at all in terms of perf.

**Comment by passuied** [comment]
@sebastienros  Interesting, my tests show that limiting CPU actually helps. Memory definitely stays under control when limiting container's memory to 256MB.
If I leave CPU unconstrained, memory does spike on startup when under load (5 concurrent users and above).
A factor is that the EC2 host has 8 virtual CPUs so not constraining the CPU on container would allow it grab as many CPUs as available, which could explain the memory jump.
Thoughts?

**Comment by passuied** [comment]
Agreed. I can confirm that too...

**Comment by effyteva** [comment]
Hi, is this expected to be fixed on .NET Core 3.0 (or any 2.2 minor release?)
We're suffering from this as well...

**Comment by Maoni0** [comment]
@effyteva would it be possible that you could try a current coreclr build?

**Comment by Rakiah** [comment]
I'm having the same problem with .NET Core 2.1.6 except setting server garbage collection to false has not effect (it raises up to 700mb for each container)

**Comment by passuied** [comment]
You may have a memory leak... you should profile your code to understand if
there's an issue there...

**Comment by PrefabPanda** [comment]
I get the memory leak even with the web API template project. If I do a Dotnet Core 2.0 one - the memory leak isn't there. It only seems to be in 2.1 and 2.2.

**Comment by Rakiah** [comment]
As PrefabPanda stated, it happens even with a template project, this is a very severe issue because 700 mb of consumption for an empty micro-service is huge

**Comment by passuied** [comment]
Do you have ServerGarbageCollection set to false in your csproj?

**Comment by PrefabPanda** [comment]
@passuied - yes, even with it set to false.

I've tried dotnet core 3.0 preview 4 (as there doesn't seem to be a docker image for preview 5) - same problem. I've set the container limits to be 512MB instead as recommended in another thread - but no luck sadly.

**Comment by Maoni0** [comment]
if you are calling GC.Collect with compacting set to true, setting memory limit or setting it to use workstation GC will not help; it just means you have that much live memory and GC cannot reclaim it. as I mentioned on the other thread, it would be helpful to understand what is taking up all this memory - there are many ways to do this when you are not running in a container but when you are, I'm not sure what tooling is available there. @sebastienros you've worked with the container scenarios, do you happen to know?

**Comment by PrefabPanda** [comment]
@Maoni - I'd love to know what's eating up the live memory - but I literally haven't got any of my own code in the project! When I first noticed the problem I was tearing large parts of my code out, then when I realised that even a plain template project had the problem - I knew it wasn't a problem I could fix anymore.

**Comment by Maoni0** [comment]
I've spoken to @sebastienros offline and he's working on this (he's also trying to repro). so he can keep this thread updated.

**Comment by sebastienros** [comment]
I am following up on the recent comment in this issue: https://github.com/aspnet/AspNetCore/issues/10200

I think the current issue should be closed as docker limits are actually not supported on 2.1 and 2.2.

**Comment by andtii** [comment]
@sebastienros Any updates on this? We also see memory beeing eaten up on .net core 2.2. We are running in linux container on aks

**Comment by jamiegs** [comment]
For people with this issue, do you have `COMPlus_PerfMapEnabled=1` or `COMPlus_EnableEventLog=1`?  I was seeing sawtooth wave looking memory usage graphs with these enabled and disabling them fixed it.  I think it's actually `COMPlus_EnableEventLog` doing it, but I disabled both so I'm not 100% sure on which.

**Comment by Maoni0** [comment]
@tommcdon is this a known issue?
